File: python/python3/Cpp/Python3LexerBase.cpp
#include "Python3Lexer.h"
#include "Python3Parser.h"
#include <regex>

using namespace antlr4;

Python3LexerBase::Python3LexerBase(antlr4::CharStream *input): Lexer(input)
{
    opened = 0;
    lastToken = nullptr;
}

void Python3LexerBase::emit(std::unique_ptr<antlr4::Token> t)
{
    tokens.push_back(cloneToken(t));
    setToken(std::move(t));
}


std::unique_ptr<antlr4::Token> Python3LexerBase::nextToken()
{
    if (_input->LA(1) == EOF && !indents.empty()) {
        for (int i = tokens.size() - 1; i >= 0; i--) {
            if (tokens[i]->getType() == EOF) {
                tokens.erase(tokens.begin() + i);
            }
        }
        emit(commonToken(Python3Parser::NEWLINE, "\n"));
        while (!indents.empty()) {
            emit(createDedent());
            indents.pop();
        }
        emit(commonToken(EOF, "<EOF>"));
    }
    std::unique_ptr<antlr4::Token> next = Lexer::nextToken();
    if (next->getChannel() == antlr4::Token::DEFAULT_CHANNEL) {
        lastToken = cloneToken(next);
    }
    if (!tokens.empty())
    {
        next = std::move(*tokens.begin());
        tokens.erase(tokens.begin());
    }
    return next;
}

std::unique_ptr<antlr4::Token> Python3LexerBase::createDedent() {
    std::unique_ptr<antlr4::CommonToken> dedent = commonToken(Python3Parser::DEDENT, "");
    return dedent;
}

std::unique_ptr<antlr4::CommonToken> Python3LexerBase::commonToken(size_t type, const std::string& text) {
    int stop = getCharIndex() - 1;
    int start = text.empty() ? stop : stop - text.size() + 1;
    return _factory->create({ this, _input }, type, text, DEFAULT_TOKEN_CHANNEL, start, stop, lastToken ? lastToken->getLine() : 0, lastToken ? lastToken->getCharPositionInLine() : 0);
}

std::unique_ptr<antlr4::CommonToken> Python3LexerBase::cloneToken(const std::unique_ptr<antlr4::Token>& source) {
    return _factory->create({ this, _input }, source->getType(), source->getText(), source->getChannel(), source->getStartIndex(), source->getStopIndex(), source->getLine(), source->getCharPositionInLine());
}

int Python3LexerBase::getIndentationCount(const std::string& spaces) {
    int count = 0;
    for (char ch : spaces) {
        switch (ch) {
            case '\t':
                count += 8 - (count % 8);
                break;
            default:
      // A normal space char.
                count++;
        }
    }

    return count;
}

bool Python3LexerBase::atStartOfInput() {
    return getCharPositionInLine() == 0 && getLine() == 1;
}

void Python3LexerBase::openBrace() {
    this->opened++;
}

void Python3LexerBase::closeBrace() {
    this->opened--;
}


void Python3LexerBase::onNewLine()
{
    std::string newLine = std::regex_replace(getText(), std::regex("[^\r\n\f]+"), "");
    std::string spaces = std::regex_replace(getText(), std::regex("[\r\n\f]+"), "");
    int next = _input->LA(1);
    int nextnext = _input->LA(2);
    if (opened > 0 || (nextnext != -1 && (next == '\r' || next == '\n' || next == '\f' || next == '#'))) {
        skip();
    }
    else {
        emit(commonToken(Python3Lexer::NEWLINE, newLine));
        int indent = getIndentationCount(spaces);
        int previous = indents.empty() ? 0 : indents.top();
        if (indent == previous) {
            skip();
        }
        else if (indent > previous) {
            indents.push(indent);
            emit(commonToken(Python3Lexer::INDENT, spaces));
        }
        else {
            while(!indents.empty() && indents.top() > indent) {
                emit(createDedent());
                indents.pop();
            }
        }
    }
}


void Python3LexerBase::reset()
{
    tokens = std::vector<std::unique_ptr<antlr4::Token>>{};
    indents = std::stack<int>{};
    opened = 0;
    lastToken = nullptr;
    Lexer::reset();
}

File: python/python3/Dart/Python3LexerBase.dart
import 'package:antlr4/antlr4.dart';
import 'dart:io';
import 'dart:core';
import 'dart:convert';
import 'dart:collection';
import 'Python3Lexer.dart';
import 'Python3Parser.dart';

abstract class Python3LexerBase extends Lexer
{
    Python3LexerBase(CharStream input) : super(input)
    {
    }

    List<Token> Tokens = [];
    List<int> Indents = [];
    int Opened = 0;
    Token? LastToken = null;


    Token MakeCommonToken(int type, String text)
    {
        int stop = this.charIndex - 1;
        int start = text.length == 0 ? stop : stop - text.length + 1;
        var ctf = this.tokenFactory;
        final t = ctf.create(
            type,
            text,
            Pair<TokenSource?, CharStream?>(this, this.inputStream),
            channel,
            start,
            charIndex - 1,
            tokenStartLine,
            tokenStartCharPositionInLine);
        return t;
    }

    @override
    void emitToken(Token token)
    {
        super.emitToken(token);
        Tokens.add(token);
    }

    Token CreateDedent()
    {
        var dedent = MakeCommonToken(Python3Parser.TOKEN_DEDENT, "");
        CommonToken t = dedent as CommonToken;
        t.line = LastToken?.line;
        return dedent;
    }

    @override Token nextToken()
    {
        // Check if the end-of-file is ahead and there are still some DEDENTS expected.
        if (this.inputStream.LA(1) == IntStream.EOF && !Indents.isEmpty)
        {
            // Remove any trailing EOF tokens from our buffer.
            for (int i = Tokens.length - 1; i >= 0; --i)
            {
                var node = Tokens[i];
                if (node.type == IntStream.EOF)
                {
                    Tokens.removeAt(i);
                }
            }
            
            // First emit an extra line break that serves as the end of the statement.
            this.emitToken(MakeCommonToken(Python3Parser.TOKEN_NEWLINE, "\n"));

            // Now emit as much DEDENT tokens as needed.
            while (! Indents.isEmpty)
            {
                this.emitToken(CreateDedent());
                Indents.removeLast();
            }

            // Put the EOF back on the token stream.
            this.emitToken(MakeCommonToken(IntStream.EOF, "<EOF>"));
        }

        Token next = super.nextToken();
        if (next.channel == Token.DEFAULT_CHANNEL)
        {
            // Keep track of the last token on the default channel.
            LastToken = next;
        }

        if (Tokens.isEmpty)
        {
            return next;
        }
        else
        {
            Token x = Tokens.first;
            Tokens.removeAt(0);
            return x;
        }
    }

    int getIndentationCount(String spaces)
    {
        int count = 0;
        for (int i = 0; i < spaces.length; i++) {
            count += spaces[i] == '\t'.codeUnitAt(0) ? 8 - (count % 8) : 1;
        }
        return count;
    }

    bool atStartOfInput()
    {
        return this.charPositionInLine == 0 && this.line == 1;
    }

    void openBrace(){
        Opened++;
    }

    void closeBrace(){
        Opened--;
    }

    void onNewLine(){
        var newLine = this.text.replaceAll(RegExp("[^\r\n\f]+"), "");
        var spaces = this.text.replaceAll(RegExp("[\r\n\f]+"), "");

        // Strip newlines inside open clauses except if we are near EOF. We keep NEWLINEs near EOF to
        // satisfy the final newline needed by the single_put rule used by the REPL.
        int? next = this.inputStream.LA(1);
        int? nextnext = this.inputStream.LA(2);
        if (Opened > 0 || (nextnext != -1 && (next == '\r'.codeUnitAt(0) || next == '\n'.codeUnitAt(0) || next == '\f'.codeUnitAt(0) || next == '#'.codeUnitAt(0))))
        {
            // If we're inside a list or on a blank line, ignore all indents, 
            // dedents and line breaks.
            skip();
        }
        else
        {
            this.emitToken(MakeCommonToken(Python3Lexer.TOKEN_NEWLINE, newLine));
            int indent = getIndentationCount(spaces);
            int previous = Indents.isEmpty ? 0 : Indents.last;
            if (indent == previous)
            {
                // skip indents of the same size as the present indent-size
                skip();
            }
            else if (indent > previous) {
                Indents.add(indent);
                this.emitToken(MakeCommonToken(Python3Parser.TOKEN_INDENT, spaces));
            }
            else {
                // Possibly emit more than 1 DEDENT token.
                while(!Indents.isEmpty && Indents.last > indent)
                {
                    this.emitToken(CreateDedent());
                    Indents.removeLast();
                }
            }
        }
    }

    @override void reset([bool resetInput = false])
    {
        Tokens = List<Token>.empty(growable: true);
        Indents = [];
        Opened = 0;
        LastToken = null;
        super.reset(resetInput);
    }
}
